# Лабораторная работа №6: Поиск экстремума функции двух переменных с помощью ГА с бинарным представлением

**Дисциплина:** Генетические и эволюционные вычисления · **Вариант 3**

**Задача:** найти минимум 7-й функции Шаффера
$$f(x_1, x_2) = (x_1^2 + x_2^2)^{0{,}25}\,\bigl[\sin^2\!\bigl(50\,(x_1^2 + x_2^2)^{0{,}1}\bigr) + 1\bigr], \qquad x_i \in [-100;\ 100],$$
глобальный минимум $f(0, 0) = 0$. Операторы варианта: **однородный кроссинговер** и **инверсия**; исследуются стратегии сокращения популяции.

**Цель работы:** модифицировать простой ГА для двумерного поиска, изучить бинарное кодирование нескольких переменных, однородный кроссинговер, оператор инверсии и стратегии формирования нового поколения.

Ноутбук рассчитан на перезапуск по ячейкам в Docker-окружении репозитория (`make up`, http://localhost:8888). Полный прогон — около 2 минут.

In [1]:
import time
from pathlib import Path

import numpy as np
import pandas as pd
import plotly
import plotly.graph_objects as go
import plotly.io as pio
from plotly.subplots import make_subplots

OUTPUTS = Path('report/outputs')
FIGURES = Path('report/figures')
TABLES = Path('report/tables')
for p in (OUTPUTS, FIGURES, TABLES):
    p.mkdir(parents=True, exist_ok=True)

# Палитра проверена на различимость при дальтонизме
PALETTE = ['#2a78d6', '#eb6834', '#1baf7a', '#eda100', '#e87ba4', '#008300', '#4a3aa7', '#e34948']
BLUE_RAMP = ['#cde2fb', '#9ec5f4', '#6da7ec', '#3987e5', '#256abf', '#184f95', '#0d366b']
INK, INK2, MUTED, GRID, AXIS, SURFACE = '#0b0b0b', '#52514e', '#898781', '#e1e0d9', '#c3c2b7', '#fcfcfb'
GOAL = '#d03b3b'

axis_style = dict(gridcolor=GRID, linecolor=AXIS, zeroline=False, ticks='outside', tickcolor=AXIS, ticklen=4,
                  title_font=dict(size=14, color=INK2), tickfont=dict(size=12, color=INK2))
pio.templates['lab'] = go.layout.Template(layout=dict(
    font=dict(family='system-ui, -apple-system, "Segoe UI", Roboto, Arial, sans-serif', size=14, color=INK),
    paper_bgcolor=SURFACE, plot_bgcolor=SURFACE, colorway=PALETTE,
    xaxis=axis_style, yaxis=axis_style,
    title=dict(x=0.01, xanchor='left', font=dict(size=18, color=INK)),
    legend=dict(bgcolor='rgba(252,252,251,0.85)', bordercolor=GRID, borderwidth=1, font=dict(size=12)),
    hoverlabel=dict(bgcolor='white', bordercolor=AXIS, font=dict(size=12, color=INK)),
    margin=dict(l=70, r=30, t=80, b=60),
))
pio.templates.default = 'plotly_white+lab'
PLOT_CONFIG = dict(displaylogo=False, responsive=True, toImageButtonOptions=dict(format='png', scale=2))


def save(fig, name, width=1000, height=520, png=True, png_fig=None):
    """HTML (интерактивный) в report/outputs и PNG (для отчёта) в report/figures."""
    fig.write_html(OUTPUTS / f'{name}.html', include_plotlyjs='cdn', config=PLOT_CONFIG, auto_play=False)
    if png:
        (png_fig or fig).write_image(FIGURES / f'{name}.png', width=width, height=height, scale=2)


def to_latex(df, name, caption, label, col_format=None):
    """Таблица результатов для отчёта: числа в отчёте совпадают с ноутбуком."""
    body = df.to_latex(index=False, escape=False, column_format=col_format or 'c' * len(df.columns))
    tex = ('\\begin{table}[H]\n\\caption{' + caption + '}\n\\label{' + label + '}\n\\centering\n\\small\n'
           + body + '\\end{table}\n')
    (TABLES / f'{name}.tex').write_text(tex, encoding='utf-8')


def anim_controls(steps, prefix='поколение: ', duration=250):
    """Кнопки «Старт / Пауза / Сброс» и слайдер для анимации Plotly."""
    play = dict(frame=dict(duration=duration, redraw=False), transition=dict(duration=int(duration * 0.8)),
                fromcurrent=True, mode='immediate')
    still = dict(frame=dict(duration=0, redraw=False), mode='immediate', transition=dict(duration=0))
    return dict(
        updatemenus=[dict(type='buttons', direction='left', x=1.0, xanchor='right', y=1.1, yanchor='bottom',
                          showactive=False, buttons=[
                              dict(label='▶ Старт', method='animate', args=[None, play]),
                              dict(label='⏸ Пауза', method='animate', args=[[None], still]),
                              dict(label='⟲ Сброс', method='animate', args=[[str(steps[0])], still])])],
        sliders=[dict(active=0, x=0, len=1, y=-0.08, yanchor='top', pad=dict(t=30),
                      currentvalue=dict(prefix=prefix, font=dict(size=13, color=INK2)),
                      steps=[dict(label=str(s), method='animate', args=[[str(s)], still]) for s in steps])])


print('numpy', np.__version__, '| pandas', pd.__version__, '| plotly', plotly.__version__)

numpy 2.4.4 | pandas 3.0.2 | plotly 6.7.0


## 1. Функция Шаффера F7

Функция многомодальна: минимумы расположены концентрическими кольцами вокруг начала координат, и чем ближе к нулю, тем кольца плотнее. Градиентный спуск застревает на первом же кольце.

In [2]:
A, B = -100.0, 100.0


def f7(x1, x2):
    r2 = x1 ** 2 + x2 ** 2
    return r2 ** 0.25 * (np.sin(50 * r2 ** 0.1) ** 2 + 1)


g_full = np.linspace(A, B, 201)
g_zoom = np.linspace(-5, 5, 201)
fig = make_subplots(rows=1, cols=2, horizontal_spacing=0.1,
                    subplot_titles=('вся область [−100; 100]²', 'окрестность минимума [−5; 5]²'))
for c, g in enumerate((g_full, g_zoom), start=1):
    X1, X2 = np.meshgrid(g, g)
    fig.add_trace(go.Contour(x=g, y=g, z=f7(X1, X2), colorscale='Blues', reversescale=True, ncontours=25,
                             showscale=c == 2, colorbar=dict(title='f', thickness=12), line=dict(width=0.3),
                             hovertemplate='x₁ = %{x:.2f}<br>x₂ = %{y:.2f}<br>f = %{z:.3f}<extra></extra>'), 1, c)
    fig.add_trace(go.Scatter(x=[0], y=[0], mode='markers', name='глобальный минимум f(0, 0) = 0', showlegend=c == 1,
                             marker=dict(symbol='star', size=15, color=PALETTE[1], line=dict(color=SURFACE, width=1.5))), 1, c)
    fig.update_xaxes(title_text='x₁', row=1, col=c)
    fig.update_yaxes(title_text='x₂', row=1, col=c, scaleanchor=f'x{c if c > 1 else ""}')
r = np.linspace(0, 100, 2000)
print(f'Значение на границе области (r = 100): f = {f7(100, 0):.2f}; на радиусе 1: {f7(1, 0):.3f}')
fig.update_layout(title='Функция Шаффера F7: линии уровня', height=520, legend=dict(orientation='h', y=-0.15))
save(fig, '01_function', width=1200, height=520)
fig.show()

Значение на границе области (r = 100): f = 10.05; на радиусе 1: 1.069


## 2. Бинарное кодирование двух переменных и предел точности

Хромосома — конкатенация двух строк по $L$ бит: $x_i = a + \mathrm{dec}(s_i)\,(b-a)/(2^L-1)$. При $L = 20$ шаг сетки $\Delta = 200/(2^{20}-1) \approx 1{,}9 \cdot 10^{-4}$.

Важная деталь: **точка 0 не представима** — $(2^L - 1)/2$ не целое, ближайшие коды дают $x = \pm\Delta/2$. Поэтому даже идеальный ГА не получит $f = 0$: нижняя граница — минимум $f$ по узлам сетки у начала координат.

In [3]:
L_BITS = 20


def decode(pop, L=L_BITS):
    w = 2 ** np.arange(L - 1, -1, -1)
    x1 = A + (pop[:, :L] @ w) * (B - A) / (2 ** L - 1)
    x2 = A + (pop[:, L:] @ w) * (B - A) / (2 ** L - 1)
    return x1, x2


def grid_floor(L, span=60):
    """Минимум f по узлам сетки L-битного кода в окрестности нуля."""
    c = (2 ** L - 1) / 2
    k = np.arange(int(c) - span, int(c) + span + 2)
    x = A + k * (B - A) / (2 ** L - 1)
    X1, X2 = np.meshgrid(x, x)
    F = f7(X1, X2)
    i = np.unravel_index(np.argmin(F), F.shape)
    return F[i], X1[i], X2[i]


df_floor = pd.DataFrame([dict(L=L, шаг=(B - A) / (2 ** L - 1), **dict(zip(['мин. f на сетке', 'x₁', 'x₂'], grid_floor(L))))
                         for L in (12, 16, 20, 24)])
F_FLOOR = grid_floor(L_BITS)[0]
print(f'Нижняя граница для L = {L_BITS}: f = {F_FLOOR:.4f}')
df_floor.style.format({'шаг': '{:.1e}', 'мин. f на сетке': '{:.4f}', 'x₁': '{:+.2e}', 'x₂': '{:+.2e}'}).hide(axis='index')

Нижняя граница для L = 20: f = 0.0200


L,шаг,мин. f на сетке,x₁,x₂
12,4.9e-02,0.2104,-2.44e-02,-2.44e-02
16,3.1e-03,0.0818,-1.53e-03,-1.53e-03
20,1.9e-04,0.0200,-9.54e-05,-9.54e-05
24,1.2e-05,0.0054,-1.79e-05,-1.79e-05


## 3. Генетический алгоритм

- **Турнирная селекция** ($k$ особей, побеждает лучшая) — не зависит от масштаба $f$, в отличие от рулетки.
- **Однородный кроссинговер** с вероятностью $P_c$: каждый бит потомка берётся от случайного родителя (маска из равновероятных 0/1).
- **Инверсия** (оператор варианта) с вероятностью $P_{inv} = 0{,}2$: случайный отрезок хромосомы переворачивается. Как и в классическом ГА Холланда, инверсия дополняет **побитовую мутацию** ($p = 1/2L$ на бит); вклад каждого оператора проверяется отдельно в разделе 4.
- **Сокращение популяции:** элитарное — из родителей и потомков остаются $N$ лучших ($\mu + \lambda$); чистая замена — родители полностью заменяются потомками; случайная замена — $N$ особей выбираются из объединения случайно.

Показатели запуска: лучшее найденное $f$; успех — $f < 0{,}1$ (точка ближе $10^{-2}$ к началу координат); **скорость** — поколение, на котором впервые достигнуто $f < 0{,}1$; доля запусков, дошедших до предела сетки $f = 0{,}020$.

In [4]:
DEFAULT = dict(pop_size=80, n_gen=200, p_c=0.85, p_inv=0.2, p_bit=1 / 40, k=3, reduction='elitist')
SUCCESS = 0.1


def run_ga(pop_size=80, n_gen=200, p_c=0.85, p_inv=0.2, p_bit=1 / 40, k=3, reduction='elitist', seed=0, track=False,
           L=L_BITS):
    rng = np.random.default_rng(seed)
    n = 2 * L
    pop = rng.integers(0, 2, (pop_size, n), dtype=np.int8)
    fit = f7(*decode(pop, L))
    best_i = int(np.argmin(fit)); best, best_f = pop[best_i].copy(), fit[best_i]
    hb, hm, snaps = [best_f], [np.median(fit)], [np.c_[decode(pop, L)]] if track else None
    hit = 0 if best_f < SUCCESS else None
    for g in range(n_gen):
        # турнир
        cand = rng.integers(0, pop_size, (pop_size, k))
        par = pop[cand[np.arange(pop_size), np.argmin(fit[cand], axis=1)]]
        # однородный кроссинговер
        p1, p2 = par[0::2], par[1::2]
        do = (rng.random(len(p1)) < p_c)[:, None]
        m = rng.integers(0, 2, p1.shape, dtype=np.int8).astype(bool) & do
        kids = np.vstack([np.where(m, p2, p1), np.where(m, p1, p2)])
        # инверсия отрезка
        for i in np.where(rng.random(len(kids)) < p_inv)[0]:
            a, b = np.sort(rng.integers(0, n, 2))
            kids[i, a:b + 1] = kids[i, a:b + 1][::-1]
        # побитовая мутация (для сравнения)
        if p_bit > 0:
            kids ^= (rng.random(kids.shape) < p_bit).astype(np.int8)
        kfit = f7(*decode(kids, L))
        if reduction == 'elitist':
            allp, allf = np.vstack([pop, kids]), np.concatenate([fit, kfit])
            keep = np.argsort(allf)[:pop_size]
        elif reduction == 'pure':
            allp, allf, keep = kids, kfit, np.arange(pop_size)
        elif reduction == 'random':
            allp, allf = np.vstack([pop, kids]), np.concatenate([fit, kfit])
            keep = rng.choice(len(allp), pop_size, replace=False)
        pop, fit = allp[keep], allf[keep]
        i = int(np.argmin(fit))
        if fit[i] < best_f:
            best, best_f = pop[i].copy(), fit[i]
        if hit is None and best_f < SUCCESS:
            hit = g + 1
        hb.append(fit.min()); hm.append(np.median(fit))
        if track:
            snaps.append(np.c_[decode(pop, L)])
    x1, x2 = decode(best[None], L)
    return dict(x=(x1[0], x2[0]), f=best_f, hist_best=np.array(hb), hist_med=np.array(hm), snaps=snaps,
                final_best=fit.min(), hit=hit)


res = run_ga(**{**DEFAULT, 'n_gen': 300}, seed=1, track=True)
print(f'Лучшая точка: x = ({res["x"][0]:+.5f}; {res["x"][1]:+.5f}), f = {res["f"]:.4f}, '
      f'расстояние до нуля {np.hypot(*res["x"]):.1e}; нижняя граница сетки {F_FLOOR:.4f}')

Лучшая точка: x = (-0.00010; +0.00010), f = 0.0200, расстояние до нуля 1.3e-04; нижняя граница сетки 0.0200


In [5]:
g = np.arange(len(res['hist_best']))
fig = go.Figure()
fig.add_trace(go.Scatter(x=g, y=res['hist_best'], mode='lines', name='лучшая особь', line=dict(color=PALETTE[0], width=2.5)))
fig.add_trace(go.Scatter(x=g, y=res['hist_med'], mode='lines', name='медиана популяции', line=dict(color=PALETTE[1], width=2)))
fig.add_hline(y=F_FLOOR, line=dict(color=GOAL, dash='dash', width=1.5), annotation_text=f'предел сетки 20 бит: {F_FLOOR:.3f}',
              annotation_position='top right', annotation_font_color=GOAL)
fig.update_layout(title='Сходимость ГА на функции Шаффера F7 (N = 80, инверсия + побитовая мутация, элитарное сокращение)', xaxis_title='поколение',
                  yaxis_title='f (log)', yaxis_type='log', hovermode='x unified', height=440,
                  legend=dict(x=0.99, xanchor='right', y=0.99))
save(fig, '02_convergence', height=440)
fig.show()

In [6]:
# Популяция на линиях уровня: поколения 0…300, масштаб подстраивается под разброс популяции
steps = [0, 2, 5, 10, 20, 40, 80, 150, 300]
Xg1, Xg2 = np.meshgrid(g_full, g_full)
fig = go.Figure()
fig.add_trace(go.Contour(x=g_full, y=g_full, z=f7(Xg1, Xg2), colorscale='Blues', reversescale=True, ncontours=25,
                         showscale=False, line=dict(width=0.3), hoverinfo='skip'))
s0 = res['snaps'][0]
fig.add_trace(go.Scatter(x=s0[:, 0], y=s0[:, 1], mode='markers', name='популяция',
                         marker=dict(color=PALETTE[1], size=8, line=dict(color=SURFACE, width=1))))
fig.frames = [go.Frame(name=str(s), data=[go.Scatter(x=res['snaps'][s][:, 0], y=res['snaps'][s][:, 1])], traces=[1],
                       layout=go.Layout(title_text=f'Популяция на линиях уровня — поколение {s}')) for s in steps]
fig.update_layout(title='Популяция на линиях уровня — поколение 0', xaxis_title='x₁', yaxis_title='x₂', width=720,
                  height=740, margin=dict(t=110, b=120), yaxis=dict(scaleanchor='x'), **anim_controls(steps))
png = make_subplots(rows=1, cols=3, horizontal_spacing=0.07,
                    subplot_titles=['поколение 0 (область ±100)', 'поколение 10 (±10)', 'поколение 300 (±0.01)'])
for c, (s, w) in enumerate([(0, 100), (10, 10), (300, 0.01)], start=1):
    gg = np.linspace(-w, w, 301); G1, G2 = np.meshgrid(gg, gg)
    png.add_trace(go.Contour(x=gg, y=gg, z=f7(G1, G2), colorscale='Blues', reversescale=True, ncontours=20,
                             showscale=False, line=dict(width=0.3)), 1, c)
    P = res['snaps'][s]
    png.add_trace(go.Scatter(x=P[:, 0], y=P[:, 1], mode='markers', showlegend=False,
                             marker=dict(color=PALETTE[1], size=7, line=dict(color=SURFACE, width=1))), 1, c)
    png.update_xaxes(range=[-w, w], row=1, col=c)
    png.update_yaxes(range=[-w, w], row=1, col=c, scaleanchor=f'x{c if c > 1 else ""}')
png.update_layout(title='Популяция сжимается к началу координат (обратите внимание на масштаб осей)', height=440)
save(fig, '03_population', width=1200, height=440, png_fig=png)
fig.show()

## 4. Стратегии сокращения и операторы мутации (20 запусков)

In [7]:
R = 20


def batch(**kw):
    rr = [run_ga(**{**DEFAULT, **kw}, seed=500 + s) for s in range(R)]
    fb = np.array([r['f'] for r in rr])
    hits = [r['hit'] for r in rr if r['hit'] is not None]
    return rr, dict(медиана_f=np.median(fb), худшее_f=fb.max(), успехов=np.mean(fb < SUCCESS) * 100,
                    до_предела=np.mean(fb <= F_FLOOR + 1e-9) * 100,
                    поколений=np.median(hits) if hits else np.nan)


tic = time.perf_counter()
red_rows, red_curves = [], {}
for red, name in [('elitist', 'элитарное (μ+λ)'), ('pure', 'чистая замена'), ('random', 'случайная замена')]:
    rr, st = batch(reduction=red)
    red_rows.append(dict(стратегия=name, **st))
    red_curves[name] = np.median([r['hist_best'] for r in rr], axis=0)
mut_rows, mut_curves = [], {}
for name, kw in [('инверсия + побитовая (по умолчанию)', dict(p_inv=0.2, p_bit=1 / 40)), ('только инверсия', dict(p_inv=0.2, p_bit=0.0)),
                 ('только побитовая', dict(p_inv=0.0, p_bit=1 / 40)), ('без мутации', dict(p_inv=0.0, p_bit=0.0))]:
    rr, st = batch(**kw)
    mut_rows.append(dict(оператор=name, **st))
    mut_curves[name] = np.median([r['hist_best'] for r in rr], axis=0)
df_red, df_mut = pd.DataFrame(red_rows), pd.DataFrame(mut_rows)
print(f'Готово за {time.perf_counter() - tic:.1f} с')
fmt = {'медиана_f': '{:.4f}', 'худшее_f': '{:.3f}', 'успехов': '{:.0f} %', 'до_предела': '{:.0f} %', 'поколений': '{:.0f}'}
display(df_red.style.format(fmt).hide(axis='index'))
df_mut.style.format(fmt).hide(axis='index')

Готово за 2.5 с


стратегия,медиана_f,худшее_f,успехов,до_предела,поколений
элитарное (μ+λ),0.0200,0.031,100 %,90 %,26
чистая замена,0.0200,0.021,100 %,65 %,44
случайная замена,0.0208,0.059,100 %,35 %,84


оператор,медиана_f,худшее_f,успехов,до_предела,поколений
инверсия + побитовая (по умолчанию),0.0200,0.031,100 %,90 %,26
только инверсия,0.0638,0.397,70 %,10 %,47
только побитовая,0.0200,0.055,100 %,80 %,29
без мутации,1.0128,2.246,5 %,0 %,11


In [8]:
fig = make_subplots(rows=1, cols=2, shared_yaxes=True, horizontal_spacing=0.04,
                    subplot_titles=('Стратегия сокращения популяции', 'Оператор мутации (элитарное сокращение)'))
for k, (name, cur) in enumerate(red_curves.items()):
    fig.add_trace(go.Scatter(x=np.arange(len(cur)), y=cur, mode='lines', name=name, legendgroup='r',
                             line=dict(color=PALETTE[k], width=2.5)), 1, 1)
for k, (name, cur) in enumerate(mut_curves.items()):
    fig.add_trace(go.Scatter(x=np.arange(len(cur)), y=cur, mode='lines', name=name, legendgroup='m',
                             line=dict(color=PALETTE[k + 3], width=2.5, dash='dot' if 'без' in name else None)), 1, 2)
for c in (1, 2):
    fig.add_hline(y=F_FLOOR, line=dict(color=GOAL, dash='dash', width=1), row=1, col=c)
    fig.update_xaxes(title_text='поколение', row=1, col=c)
fig.update_yaxes(type='log')
fig.update_yaxes(title_text='лучшее f поколения (медиана 20 запусков)', row=1, col=1)
fig.update_layout(title='Лучшая особь текущего поколения', height=470, legend=dict(orientation='h', y=-0.2, x=0))
save(fig, '04_reduction_mutation', width=1300, height=470)
fig.show()

## 5. Влияние параметров (20 запусков на значение)

In [9]:
GRID = {'pop_size': [20, 40, 80, 160], 'p_c': [0.0, 0.5, 0.85, 1.0], 'p_inv': [0.0, 0.05, 0.2, 0.5], 'k': [2, 3, 5, 8]}
LABELS = {'pop_size': 'размер популяции N', 'p_c': 'вероятность кроссинговера Pc', 'p_inv': 'вероятность инверсии Pinv',
          'k': 'размер турнира k'}
rows = []
tic = time.perf_counter()
for par, vals in GRID.items():
    for v in vals:
        rows.append(dict(параметр=LABELS[par], par=par, значение=v, **batch(**{par: v})[1]))
df_par = pd.DataFrame(rows)
print(f'Готово за {time.perf_counter() - tic:.1f} с')
df_par.drop(columns='par').style.format(fmt).hide(axis='index')

Готово за 6.6 с


параметр,значение,медиана_f,худшее_f,успехов,до_предела,поколений
размер популяции N,20.000000,0.0200,0.240,80 %,70 %,58
размер популяции N,40.000000,0.0200,0.087,100 %,70 %,38
размер популяции N,80.000000,0.0200,0.031,100 %,90 %,26
размер популяции N,160.000000,0.0200,0.020,100 %,100 %,22
вероятность кроссинговера Pc,0.000000,0.0200,0.055,100 %,80 %,36
вероятность кроссинговера Pc,0.500000,0.0200,0.031,100 %,85 %,32
вероятность кроссинговера Pc,0.850000,0.0200,0.031,100 %,90 %,26
вероятность кроссинговера Pc,1.000000,0.0200,0.031,100 %,85 %,24
вероятность инверсии Pinv,0.000000,0.0200,0.055,100 %,80 %,29
вероятность инверсии Pinv,0.050000,0.0200,0.031,100 %,95 %,27


In [10]:
fig = make_subplots(rows=1, cols=4, shared_yaxes=True, horizontal_spacing=0.03, subplot_titles=[LABELS[p] for p in GRID])
for c, par in enumerate(GRID, start=1):
    d = df_par[df_par.par == par]
    fig.add_trace(go.Bar(x=[f'{v:g}' for v in d.значение], y=d.поколений, showlegend=False,
                         marker=dict(color=PALETTE[0], line=dict(color=SURFACE, width=2)),
                         text=[f'{v:.0f}' for v in d.поколений], textposition='outside', textfont=dict(size=11, color=INK2),
                         customdata=np.c_[d.успехов, d.худшее_f],
                         hovertemplate='%{x}: поколений до f<0.1 — %{y:.0f}<br>успехов %{customdata[0]:.0f}%<br>'
                                       'худшее f = %{customdata[1]:.3f}<extra></extra>'), 1, c)
fig.update_xaxes(type='category')
fig.update_yaxes(title_text='поколений до f < 0.1 (медиана)', row=1, col=1)
fig.update_layout(title='Влияние параметров на скорость поиска: поколений до f < 0.1 (медиана 20 запусков)',
                  height=440, bargap=0.3)
save(fig, '05_parameters', width=1300, height=440)
fig.show()

## 6. Таблицы для отчёта

In [11]:
def f4(v):
    return f'{v:.4f}'


t1 = pd.DataFrame({'Показатель': ['Лучшая точка $(x_1; x_2)$', 'Расстояние до $(0; 0)$', 'Лучшее $f$ (ГА)',
                                  'Нижняя граница сетки 20 бит', 'Вычислений функции'],
                   'Значение': [f'({res["x"][0]:+.2e}; {res["x"][1]:+.2e})', f'{np.hypot(*res["x"]):.1e}', f4(res['f']),
                                f4(F_FLOOR), f'{80 * 301}']})
to_latex(t1, 'tab_result', 'Результат основного запуска (N = 80, 300 поколений, инверсия и побитовая мутация, элитарное сокращение)', 'tab:result', 'lc')


def tab(df, first, name, caption, label):
    t = pd.DataFrame({first.capitalize(): df[first].str.replace('μ+λ', '$\\mu+\\lambda$', regex=False).str.replace(' (по умолчанию)', '', regex=False), 'Медиана $f$': df.медиана_f.map(f4),
                      'Худшее $f$': df.худшее_f.map(lambda v: f'{v:.3f}'),
                      'Успехов, \\%': df.успехов.map(lambda v: f'{v:.0f}'), 'До предела, \\%': df.до_предела.map(lambda v: f'{v:.0f}'),
                      'Поколений': df.поколений.map(lambda v: '---' if np.isnan(v) else f'{v:.0f}')})
    to_latex(t, name, caption, label, 'l' + 'c' * 5)


tab(df_red, 'стратегия', 'tab_reduction', 'Стратегии сокращения популяции (20 запусков)', 'tab:reduction')
tab(df_mut, 'оператор', 'tab_mutation', 'Вклад операторов мутации (20 запусков, элитарное сокращение)', 'tab:mutation')
t4 = pd.DataFrame({'Параметр': df_par.параметр.str.replace('Pc', '$P_c$').str.replace('Pinv', '$P_{inv}$'),
                   'Значение': df_par.значение.map(lambda v: f'{v:g}'), 'Худшее $f$': df_par.худшее_f.map(lambda v: f'{v:.3f}'),
                   'Успехов, \\%': df_par.успехов.map(lambda v: f'{v:.0f}'), 'До предела, \\%': df_par.до_предела.map(lambda v: f'{v:.0f}'),
                   'Поколений': df_par.поколений.map(lambda v: '---' if np.isnan(v) else f'{v:.0f}')})
to_latex(t4, 'tab_params', 'Влияние параметров ГА (20 запусков на значение)', 'tab:params', 'lccccc')
print(sorted(p.name for p in TABLES.glob('*.tex')))

['tab_mutation.tex', 'tab_params.tex', 'tab_reduction.tex', 'tab_result.tex']


## Выводы

1. **Результат.** ГА (N = 80, однородный кроссинговер, инверсия + побитовая мутация, элитарное сокращение) находит точку $(-1{,}0 \cdot 10^{-4};\ 1{,}0 \cdot 10^{-4})$ с $f = 0{,}0200$ — это **точный предел 20-битного кода**: ноль не представим, ближайшие узлы сетки отстоят от него на $\Delta/2$, и меньшего $f$ на этой сетке не существует. Популяция за 10 поколений стягивается с области $\pm 100$ к кольцам радиусом несколько единиц, к 40-му поколению достигает предела.
2. **Надёжность.** В 20 запусках успех ($f < 0{,}1$) — 100 %, предел сетки достигнут в 90 % запусков, медиана — 26 поколений до $f < 0{,}1$.
3. **Инверсия без побитовой мутации слаба:** 70 % успехов и лишь 10 % запусков до предела. Инверсия только переставляет биты — набор нулей и единиц в хромосоме не меняется, и недостающие значения битов не появляются. Побитовая мутация одна даёт 100 % / 80 %, вместе с инверсией — 100 % / 90 % и лучшее худшее значение (0,031). Без мутации популяция вырождается: 5 % успехов.
4. **Стратегия сокращения** определяет скорость: элитарная ($\mu + \lambda$) — 26 поколений и 90 % до предела, чистая замена — 44 поколения и 65 %, случайная замена — 84 поколения и 35 %. Без элитизма лучшие особи теряются, и кривая лучшей особи поколения колеблется.
5. **Параметры:** рост популяции с 20 до 160 ускоряет поиск (58 → 22 поколения) и поднимает долю запусков до предела с 70 до 100 %. Увеличение размера турнира ускоряет сходимость (31 → 20 поколений при k = 2 → 8), но при большом давлении отбора повышается риск застрять на кольце. Кроссинговер ускоряет поиск (36 → 24 поколения при Pc = 0 → 1).